packages/runtime/container-runtime/src/runningSummarizer.ts

```diff
@@ -353,41 +438,35 @@ export class RunningSummarizer implements IDisposable {
 		this.initialized = true;
 	}
 
-	/**
-	 * Blocks a new summarizer from running in case RefreshSummaryAck is being processed.
-	 * Assumes that caller checked upfront for lack of concurrent action (this.refreshSummaryAckLock)
-	 * before calling this API. I.e. caller is responsible for either erroring out or waiting on this promise.
-	 * Note: The refreshSummaryAckLock makes sure no summarizer gets enqueued or processed
-	 * until the refresh has completed. One can't rely uniquely on the summarizingLock as the
-	 * refreshLatestSummaryAck also happens during the time summarizingLock !== undefined.
-	 * Ex. Summarizer submits a summay + op and then waits for the Summary Ack to proceed
-	 * with the refreshLatestSummaryAck and complete the summary.
-	 * @param action - action to perform.
-	 * @returns - result of action.
-	 */
-	public async lockedRefreshSummaryAckAction<T>(action: () => Promise<T>) {
-		assert(
-			this.refreshSummaryAckLock === undefined,
-			0x396 /* Refresh Summary Ack - Caller is responsible for checking lock */,
-		);
+	private beforeSummaryAction() {
+		this.summarizeCount++;
+	}
 
-		const refreshSummaryAckLock = new Deferred<void>();
-		this.refreshSummaryAckLock = refreshSummaryAckLock.promise;
+	private afterSummaryAction() {
+		const retry = this.tryWhileSummarizing;
+		this.tryWhileSummarizing = false;
 
-		return action().finally(() => {
-			refreshSummaryAckLock.resolve();
-			this.refreshSummaryAckLock = undefined;
-		});
+		// After summarizing, we should check to see if we need to summarize again.
+		// Rerun the heuristics and check for enqueued summaries.
+		if (!this.stopping && !this.tryRunEnqueuedSummary() && retry) {
+			this.heuristicRunner?.run();
+		}
 	}
 
 	/**
 	 * Runs single summary action that prevents any other concurrent actions.
 	 * Assumes that caller checked upfront for lack of concurrent action (this.summarizingLock)
 	 * before calling this API. I.e. caller is responsible for either erroring out or waiting on this promise.
 	 * @param action - action to perform.
+	 * @param before - set of instructions to run before running the action.
+	 * @param after - set of instructions to run after running the action.
 	 * @returns - result of action.
 	 */
-	private async lockedSummaryAction<T>(action: () => Promise<T>) {
+	private async lockedSummaryAction<T>(
+		action: () => Promise<T>,
```

Comment:
nit: Re-order the params to: before, action and after so it is more inline with how these will be processed.
Another option is to move action to the end (before, after, action) because the action can be a large block (as in trySumamrize) making it a little difficult to read what before and after are doing.